### ISSUE TITLE
Design of TabNet

### ISSUE THREAD

**Issue by hosseinghafarian** [dreamquark-ai/tabnet#387]
**Describe the bug**

Actually, it is not a bug. In my opinion a minor design issue. According to S.O.L.I.D. object oriented design principles, it seems that the TabModel class which is an abstract class must not be dependent to TabNet. Moreover, I don't understand why self.network, returned a negative M_loss and later in "loss = loss - self.lambda_sparse * M_loss", it uses its minus, in def _train_batch(self, X, y) method. Isn't it better to send a positive M_loss? Is there any reason for this? 

**What is the current behavior?**
Describe above
**If the current behavior is a bug, please provide the steps to reproduce.**

**Expected behavior**

It is much better to change TabModel class so that it is not dependent on TabNet class and its properties, such as n_a, n_d, etc. 
**Screenshots**

**Other relevant information:**
poetry version:  
python version: 
Operating System: 
Additional tools:

**Additional context**

**Comment by Optimox** [comment]
Hello @hosseinghafarian,

About the minus sign before the loss, I guess it's just mathematically correct. It might have been clearer to return a positive loss and sum them, but we are talking about a minus sign here, no big deal.

About the S.O.L.I.D. object oriented design principles, I had to search for the term on the internet to be honest. There may be conceptual flaws in the architecture of the repo, I don't deny it. If you are willing to make a PR to propose something better please do.

**Comment by Optimox** [comment]
@hosseinghafarian,

I'm closing this, feel free to reopen as soon as you have a proposition of change that we can discuss.
